pallets/loans-ref/src/lib.rs

```diff
@@ -679,16 +681,38 @@ pub mod pallet {
 			})
 		}
 
-		fn find_write_off_state(
+		/// From all overdue write off rules, it returns the one has highest percentage
+		/// (or highest penalty, if same percentage) that can be applied.
+		///
+		/// Suppose a policy with the following rules:
+		/// - overdue_days: 5,   percentage 10%
+		/// - overdue_days: 10,  percentage 30%
+		/// - overdue_days: 15,  percentage 20%
+		///
+		/// If the loan is not overdue, it will not return any rule.
+		/// If the loan overdue by 4 days, it will not return any rule.
+		/// If the loan is overdue by 9 days, it will return the first rule.
+		/// If the loan is overdue by 60 days, it will return the second rule
+		/// (because it has a higher percetage).
+		fn find_write_off_rule(
 			pool_id: PoolIdOf<T>,
-			maturity_date: Moment,
-		) -> Result<WriteOffState<T::Rate>, DispatchError> {
-			WriteOffState::find_best(
-				WriteOffPolicy::<T>::get(pool_id).into_iter(),
-				maturity_date,
-				T::Time::now().as_secs(),
-			)
-			.ok_or_else(|| Error::<T>::NoValidWriteOffState.into())
+			loan: &ActiveLoan<T>,
+		) -> Result<Option<WriteOffRule<T::Rate>>, DispatchError> {
+			Ok(WriteOffPolicy::<T>::get(pool_id)
+				.into_iter()
+				.filter_map(|rule| {
+					rule.triggers
+						.iter()
+						.map(|trigger| loan.check_write_off_trigger(&trigger))
+						.find(|e| match e {
+							Ok(value) => *value,
+							Err(_) => true,
+						})
+						.map(|result| result.map(|_| rule))
+				})
+				.collect::<Result<Vec<_>, _>>()? // This exits if error before getting the maximum
+				.into_iter()
+				.max_by(|r1, r2| r1.status.cmp(&r2.status)))
```

Comment:
Maybe it's not trivial to understand fastly. This is mostly because `Iter::try_find()` is unstable, and the workaround is easy to see. With `try_find` this should be 3-4 lines 😢 

Basically, we want to accept a "trigger" that has been true or with an error, to be able to catch that error successfully later and not avoid it without noise.